docs/modeling.md

```diff
@@ -1,67 +1,684 @@
+```
+
+This requires knowledge of the API of the underlying solver, as any function name that you give to it will be called. The only special thing that the DirectConstraint does, is automatically translate any CPMpy variable in the argument to the native solver variable.
+
+Note that any argument given will be checked for whether it needs to be mapped to a native solver variable. This may give errors on complex arguments, or be inefficient. You can tell the `DirectConstraint` not to scan for variables with `noarg` argument, for example:
+
+```python
+from cpmpy import *
+trans_vars = boolvar(shape=4, name="trans")
+
+s = SolverLookup.get("ortools")
+
+trans_tabl = [ # corresponds to regex 0* 1+ 0+
+    (0, 0, 0),
+    (0, 1, 1),
+    (1, 1, 1),
+    (1, 0, 2),
+    (2, 0, 2)
+]
+s += DirectConstraint("AddAutomaton", (trans_vars, 0, [2], trans_tabl),
+                      novar=[1, 2, 3])  # optional, what not to scan for vars
+```
+
+A minimal example of the DirectConstraint for every supported solver is [in the test suite](https://github.com/CPMpy/cpmpy/tree/master/tests/test_direct.py).
+
+The `DirectConstraint` is a very powerful primitive to get the most out of specific solvers. See the following examples: [nonogram_ortools.ipynb](https://github.com/CPMpy/cpmpy/tree/master/examples/nonogram_ortools.ipynb) using of a helper function that generates automatons with DirectConstraints; [vrp_ortools.py](https://github.com/CPMpy/cpmpy/tree/master/examples/vrp_ortools.ipynb) demonstrating ortools' newly introduced multi-circuit global constraint through DirectConstraint; and [pctsp_ortools.py](https://github.com/CPMpy/cpmpy/tree/master/examples/pctsp_ortools.ipynb) that uses a DirectConstraint to use ortools circuit to post a sub-circuit constraint as needed for this price-collecting TSP variant.
+
+### Directly accessing the underlying solver
+
+The `DirectConstraint("AddAllDifferent", iv)` is equivalent to the following code, which demonstrates that you can mix the use of CPMpy with calling the underlying solver directly: 
+
+```python
+from cpmpy import *
+iv = intvar(1,9, shape=3)
+
+s = SolverLookup.get("ortools")
+
+s += AllDifferent(iv)  # the traditional way, equivalent to:
+s.ort_model.AddAllDifferent(s.solver_vars(iv))  # directly calling the API, has to be with native variables
+```
+
+observe how we first map the CPMpy variables to native variables by calling `s.solver_vars()`, and then give these to the native solver API directly.  This is in fact what happens behind the scenes when posting a DirectConstraint, or any CPMpy constraint.
+
+While directly calling the solver offers a lot of freedom, it is a bit more cumbersome as you have to map the variables manually each time. Also, you no longer have a declarative model that you can pass along, print or inspect. In contrast, a `DirectConstraint` is a CPMpy expression so its use is identical to other constraints.
```

Comment:
Well except that you cannot use it in a nested context?
I.e, I don't think you can do `bv.implies(DirectConstraint("AddAllDifferent", iv)`